service/interpreter/workflowImpl.go

```diff
@@ -808,6 +806,38 @@ func invokeStateExecute(
 	return &decision, service.CompletedStateExecutionStatus, nil
 }
 
+func failOnSignalWithStart(state iwfidl.StateMovement, provider WorkflowProvider, err error) (*iwfidl.StateDecision, service.StateExecutionStatus, error) {
+	if shouldProceedOnExecuteApiError(state) {
+		return nil, service.ExecuteApiFailedAndProceed, nil
+	}
+	return nil, service.FailureStateExecutionStatus, convertStateApiActivityError(provider, err)
+}
+
+func signalWithStartWorkflowId(state iwfidl.StateMovement, parentId string, executionContext iwfidl.Context) string {
+	if state.WaitForKey != nil {
+		return service.IwfSystemConstPrefix + parentId + "_" + state.StateId + "_" + *state.WaitForKey
+	} else {
+		return service.IwfSystemConstPrefix + parentId + "_" + *executionContext.StateExecutionId
+	}
```

Comment:
Nit: it would be nice to move this to a common place to share with api service: `GetWorkflowIdForWaitForStateExecution(parentId, stateExeId, waitForKey, stateId)`

Reply:
Done